File: app/src/main/java/org/dhis2/data/server/ConfigurationRepositoryImpl.java
package org.dhis2.data.server;

import androidx.annotation.NonNull;

import org.hisp.dhis.android.core.configuration.Configuration;
import org.hisp.dhis.android.core.configuration.ConfigurationManager;

import io.reactivex.Observable;
import okhttp3.HttpUrl;

public class ConfigurationRepositoryImpl implements ConfigurationRepository {
    @NonNull
    private final ConfigurationManager configurationManager;

    public ConfigurationRepositoryImpl(@NonNull ConfigurationManager configurationManager) {
        this.configurationManager = configurationManager;
    }

    @NonNull
    @Override
    public Observable<Configuration> configure(@NonNull HttpUrl baseUrl) {
        return Observable.defer(() -> Observable.fromCallable(
                () -> {
                    Configuration configuration = Configuration.builder().serverUrl(baseUrl).build();
                    configurationManager.configure(configuration);
                    return configuration;
                }));
    }

    @NonNull
    @Override
    public Observable<Configuration> get() {
        return Observable.defer(() -> {
            Configuration configuration = configurationManager.get();
            if (configuration != null) {
                return Observable.just(configuration);
            }

            return Observable.empty();
        });
    }

    @NonNull
    @Override
    public Observable<Integer> remove() {
        return Observable.defer(() -> Observable.fromCallable(configurationManager::remove));
    }
}

File: app/src/main/java/org/dhis2/data/server/ConfigurationRepositoryImpl.kt
package org.dhis2.data.server

import io.reactivex.Observable
import okhttp3.HttpUrl
import org.hisp.dhis.android.core.configuration.Configuration
import org.hisp.dhis.android.core.configuration.ConfigurationManager
import java.util.concurrent.Callable

class ConfigurationRepositoryImpl(
        val configurationManager: ConfigurationManager
): ConfigurationRepository {
    override fun configure(baseUrl: HttpUrl): Observable<Configuration> {
        return Observable.defer {
            Observable.fromCallable {
                val configuration = Configuration.builder().serverUrl(baseUrl).build()
                configurationManager.configure(configuration)
                configuration
            }
        }
    }

    override fun get(): Observable<Configuration> {
        return Observable.defer {
            val configuration = configurationManager.get()
            return@defer if (configuration != null) {
                Observable.just(configuration)
            } else Observable.empty()
        }
    }

    override fun remove(): Observable<Int> {
        return Observable.defer { Observable.fromCallable { configurationManager.remove() } }
    }

}